### PR TITLE
Enable and expose Schnorrsig

### PR DESCRIPTION
**Pull request by Gruruya**
This patchset enables and adds an interface for multikey and schnorrsig in `abi.nim`.

I'm not sure how you would want the interface in `secp256k1.nim` to look, so I haven't touched that.

### CODE REVIEW

File: secp256k1/abi.nim [hunk 1]
```diff
@@ -223,7 +223,7 @@ var secp256k1_nonce_function_default*: secp256k1_nonce_function
 ## Recovery interface follows
 
 type
-  secp256k1_ecdsa_recoverable_signature* = object
+  secp256k1_ecdsa_recoverable_signature* {.bycopy.} = object
```
Comment (arnetheduck) [inline secp256k1/abi.nim:226]: why the `bycopy` here?

**Comment by arnetheduck** [comment]
Thanks for the PR!

> I'm not sure how you would want the interface

Generally, the interface strives to provide a "safe" experience for Nim users, as far as the language allows one to do so - it's fine that we start with the ABI - my suggestion would be that you start using the raw ABI and think about how to safely wrap it in your application - once you have something you're satisfied with, we can revisit.

File: secp256k1/abi.nim (hunk 1 above)
Comment (Gruruya) [reply secp256k1/abi.nim:226]: Oh, you're right. Looks like it's only ever passed with a pointer. Shouldn't need `bycopy` then.

**Comment by Gruruya** [comment]
`secp256k1_keypair_create` requires a context with precomputed tables, so I can't use `secp256k1_context_no_precomp`.

If I'm replacing SkKeyPair with an object containing `secp256k1_keypair`, I'll have to rework `random`.
I'm thinking of making `random(SkKeyPair, rng)` into `random(SkKeyPair, rng, ctx = getContext())` to keep compatability, but computing the context each time could harm performance.

I could also make a new type instead of replacing `SkKeypair`, such as `SkMultiKeypair`, so that `ctx` could be a required parameter and the performance of `SkKeyPair` remains unaffected.

Which one do you like more?

**Comment by Gruruya** [comment]
Currently, I'm rewriting SkKeyPair to use `secp256k1_keypair` so that I can use that for Schnorr signatures.

I've been able to mostly keep the behavior of `keypair.pubkey`, `keypair.seckey`, and `keypair.seckey =` the same. However, the field accesses are copies, `seckey =` changes the `pubkey` as well, and `pubkey =` is a no-op.

Most code would be unaffected by these changes, however, `clear keypair.seckey` now errors, instead requiring `clear keypair`. Would this breaking change be acceptable?

If not, I could instead make a new type such as `SkMultiKeyPair`, or I could make an implicit `secp256k1_keypair` out of the old `SkKeyPair` every time I need a `secp256k1_keypair`. What do you think?

**Comment by Gruruya** [comment]
Good to go

**Review by zah — APPROVED** [review]

File: secp256k1/abi.nim [hunk 2]
```diff
@@ -326,3 +326,200 @@ proc secp256k1_ecdh_raw*(ctx: ptr secp256k1_context; output32: ptr byte;
...
+                                  keypair: ptr secp256k1_keypair): cint {.secp.}
+
+proc secp256k1_keypair_xonly_tweak_add*(ctx: ptr secp256k1_context;
+                                        keypair: ptr secp256k1_keypair;
+                                        tweak32: ptr byte): cint {.secp.}
+
+## Schnorrsig interface follows
+
+type
+  secp256k1_nonce_function_hardened* {.bycopy.} = object
```
Comment (arnetheduck) [inline secp256k1/abi.nim:425]:
```suggestion
  secp256k1_nonce_function_hardened* = object
```

**Comment by arnetheduck** [comment]
>  32-bit array of (presumably) random bytes?

hm, with the ability to pass in an RNG, this is strictly not needed (the rng can simply return a given array), but I do wonder about the convenience.

Also, compared to the C version, we're making the "random-less" version easier to call (there, you need to explicitly specify `null`) - an explicit "no-randomness" version might be preferable, whether that is through `Opt[Rng]` or some other convention.

**Comment by arnetheduck** [comment]
>  Its using secp256k1_ec_seckey_verify to check the RNG, but I'm not sure if this is excessive. Is there a more suitable way to check the RNG?

Per https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#default-signing, the randomness is just that: random bytes - there is no need for further validation (unlike private keys which have a "valid range" and the randomness is actually treated as a number)

File: secp256k1.nim [hunk 3]
```diff
@@ -445,10 +474,99 @@ func signRecoverable*(key: SkSecretKey, msg: SkMessage): SkRecoverableSignature
...
+func signSchnorr*(key: SkSecretKey, msg: openArray[byte]): SkSchnorrSignature =
+  ## Sign message `msg` using private key `key` with the Schnorr signature algorithm and return signature object.
+  signSchnorrImpl(
+    secp256k1_schnorrsig_sign_custom(
+      getContext(), data.baseAddr, msg.baseAddr, csize_t msg.len, addr kp, nil))
+
+template signSchnorrRngImpl(makeKeypair: varargs[untyped]): untyped {.dirty.} =
+  var randbytes: array[32, byte]
+  while rng(randbytes):
+    if secp256k1_ec_seckey_verify(secp256k1_context_no_precomp, randbytes.baseAddr) == 1:
```
[the remark this answers is not in the document]
Comment (Gruruya) [reply secp256k1.nim:503]: Alright, I'll fix it. I wasn't sure what `rng` could be and if running it once would be good enough.

File: secp256k1/abi.nim (hunk 2 above)
Comment (Gruruya) [reply secp256k1/abi.nim:425]: Ah, thanks.

**Comment by Gruruya** [comment]
> Also, compared to the C version, we're making the "random-less" version easier to call (there, you need to explicitly specify `null`) - an explicit "no-randomness" version might be preferable, whether that is through `Opt[Rng]` or some other convention.

I've added a comment. Could also issue a warning or require the user to specify null somehow.
Should I remove `signSchnorr` without an `rng` from tests to limit its visbility?

**Comment by Gruruya** [comment]
I'm getting an error `cannot open file: stew/byteutils` on devel, should I include a nimble lock update with this pr? Also a version bump would be good.

**Comment by arnetheduck** [comment]
> Removing the nimble.lock file fixes this, but updating it doesn't.

Remove it - it's not yet ready for prime time

**Comment by Gruruya** [comment]
I'm using this patchset now without anything on top, it should be ready.

**Review by arnetheduck — APPROVED** [review]

**Pull request merged**